crates/walrus-sui/src/client/retry_client/retriable_sui_client.rs

```diff
@@ -183,22 +194,73 @@ impl RetriableSuiClient {
         amount: u128,
         exclude: Vec<ObjectID>,
     ) -> SuiRpcResult<Vec<Coin>> {
-        let mut total = 0u128;
-        let coins = self
+        let mut coins_stream = pin!(self
             .get_coins_stream_retry(address, coin_type)
-            .filter(|coin: &Coin| future::ready(!exclude.contains(&coin.coin_object_id)))
-            .take_while(|coin: &Coin| {
-                let ready = future::ready(total < amount);
-                total += coin.balance as u128;
-                ready
-            })
-            .collect::<Vec<_>>()
-            .await;
+            .filter(|coin: &Coin| future::ready(!exclude.contains(&coin.coin_object_id))));
+
+        // Use a binary heap to keep track of the coins with the largest balances.
+        // `Reverse` turns the max heap into a min heap.
+        let mut selected_coins =
+            BinaryHeap::<Reverse<OrderedCoin>>::with_capacity(MAX_GAS_PAYMENT_OBJECTS);
+        let mut total_selected = 0u128;
+        let mut total_available = 0u128;
+
+        while let Some(coin) = coins_stream.as_mut().next().await {
+            let coin_balance = coin.balance as u128;
```

Comment:
Nit: Can you use `into` or `from` to emphasize that this conversion always succeeds? (Also below.)